### PR TITLE
tpm_verify_device: New cases to verify TPM devices

### PR DESCRIPTION
**Pull request by YongxueHong**
ID: 1861265, 1861271, 1861273, 1861274

Signed-off-by: Yongxue Hong <yhong@redhat.com>

### CODE REVIEW

**Comment by YongxueHong** [comment]
Depends on https://github.com/avocado-framework/avocado-vt/pull/2673

**Comment by YongxueHong** [comment]
hi @vivianQizhu @qcheng-redhat
Could you help me to review, thanks.
cc @CongLi

**Comment by qcheng-redhat** [comment]
Suggest to add check if swtpm command or package exists on the system before start test case.

File: qemu/tests/tpm_verify_device.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+from virttest import error_context
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Verify the TPM device info inside guest and host.
+    Steps:
+        1. Boot guest with a emulator TPM device or pass through device.
```
Comment (qcheng-redhat) [inline qemu/tests/tpm_verify_device.py:14]: Suggest change to:
Boot guest with a emulated TPM device or passthrough TPM device.

File: qemu/tests/tpm_verify_device.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        2. Check then verify the TPM info inside guest.
```
Comment (qcheng-redhat) [inline qemu/tests/tpm_verify_device.py:15]: Suggest change to:
2. Check and verify TPM/vTPM device info  inside guest.

File: qemu/tests/tpm_verify_device.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        3. Check then verify the TPM of the OVMF log on the host.
```
Comment (qcheng-redhat) [inline qemu/tests/tpm_verify_device.py:16]: Suggest change to: 
3. Check and verify TPM/vTPM device info in the OVMF log on host.

**Comment by qcheng-redhat** [comment]
Need add 
only q35 
only ovmf 
in .cfg for x86 host.

File: qemu/tests/tpm_verify_device.py (hunk 1 above)
Comment (YongxueHong) [reply qemu/tests/tpm_verify_device.py:14]: Thanks, have updated.

File: qemu/tests/tpm_verify_device.py (hunk 3 above)
Comment (YongxueHong) [reply qemu/tests/tpm_verify_device.py]: Thanks, have updated.

**Comment by YongxueHong** [comment]
> Need add
> only q35
> only ovmf
> in .cfg for x86 host.

Thanks, have updated.

**Comment by YongxueHong** [comment]
Hi @ngu-niny 
Please help me to review, thanks.

**Comment by qcheng-redhat** [comment]
Get error msg:

ERROR: Command 'dmesg | grep ' TPM ' | awk '{print $5}'' failed.\nstdout: b''\nstderr: b"dmesg: bad usage\nTry 'dmesg --help' for more information.\n"\nadditional_info: None

**Comment by YongxueHong** [comment]
> Get error msg:
> 
> ERROR: Command 'dmesg | grep ' TPM ' | awk '{print $5}'' failed.\nstdout: b''\nstderr: b"dmesg: bad usage\nTry 'dmesg --help' for more information.\n"\nadditional_info: None

Hi @qcheng-redhat 
My fault, fix it, please help me to review it again, thanks.

File: qemu/tests/tpm_verify_device.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        for pattern in patterns.split(split_string):
+            if not re.search(r'%s' % pattern, string, flags):
+                test.fail('No Found pattern "%s" from "%s".' % (pattern, string))
+            if re.search(r'error', string, re.M | re.A):
+                test.error('Found errors from "%s".' % string)
+
+    cmd_get_tpm_ver = params.get('cmd_get_tpm_ver')
+    if cmd_get_tpm_ver:
+        actual_tpm_ver = process.system_output(cmd_get_tpm_ver,
+                                               shell=False).decode().strip()
```
Comment (qcheng-redhat) [inline qemu/tests/tpm_verify_device.py:34]: shell=True

Comment (YongxueHong) [reply qemu/tests/tpm_verify_device.py]: have update.

**Comment by qcheng-redhat** [comment]
Passthrough case with Linux guest, got error:
FAIL: No Found pattern "/dev/tpmrm0" from "crw-rw----. 1 tss root 10, 224 Aug 26 18:43 /dev/tpm0\n".

**Comment by YongxueHong** [comment]
> Passthrough case with Linux guest, got error:
> FAIL: No Found pattern "/dev/tpmrm0" from "crw-rw----. 1 tss root 10, 224 Aug 26 18:43 /dev/tpm0\n".

Hi @qcheng-redhat 
Have fix it, pls help me to review again, thanks.

**Comment by ngu-niny** [comment]
Test PASS on ppc:

 (1/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.tpm_verify_device.with_emulator: PASS (120.21 s)
 (2/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot: PASS (1769.59 s)
 (3/4) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.ppc64le.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms: PASS (419.35 s)

Please note when using venv on some power9 host, if run the test more than once, the test might be a failure due to unable to connect '*swtpm.sock', while turnning to test manually is a PASS. It should relate to the avocado test environment. Anyway, I can accept the patch if not taking venv.

Acked-by: Nini Gu <ngu@redhat.com>

**Comment by qcheng-redhat** [comment]
Windows guest:
python3 ConfigTest.py --platform=x86_64 --guestname=Win2019,Win10,Win2016 --testcase=tpm_unattended_install,tpm_verify_device.with_emulator --machine=q35 --firmware=ovmf --customsparams=cdrom_virtio=isos/windows/virtio-win-1.9.12-2.el8.iso

(01/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: PASS (107.94 s)
 (02/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: PASS (93.61 s)
 (03/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: PASS (981.17 s)
 (04/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: PASS (242.81 s)
 (05/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.tpm_unattended_install.q35: PASS (4131.54 s)
 (06/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: PASS (56.71 s)
 (07/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: PASS (55.59 s)
 (08/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: PASS (730.93 s)
 (09/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: PASS (173.13 s)
 (10/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.tpm_unattended_install.q35: PASS (627.77 s)
 (11/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: PASS (134.02 s)
 (12/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: PASS (87.68 s)
 (13/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: PASS (823.60 s)
 (14/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: PASS (221.91 s)
 (15/15) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.tpm_unattended_install.q35: PASS (991.58 s)

Linux guest: 
python3 ConfigTest.py --platform=x86_64 --guestname=RHEL.8.2.1,RHEL8.3.0 --testcase=tpm_verify_device --machine=q35 --firmware=ovmf --clone=no
 (1/5) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.1.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.q35: PASS (49.17 s)
 (2/5) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.1.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_check_ovmf.q35: PASS (38.35 s)
 (3/5) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.1.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_reboot.q35: PASS (603.24 s)
 (4/5) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.1.x86_64.io-github-autotest-qemu.tpm_verify_device.with_emulator.with_multi_vms.q35: PASS (80.10 s)
 (5/5) Host_RHEL.m8.u3.product_av.ovmf.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.1.x86_64.io-github-autotest-qemu.tpm_verify_device.with_passthrough.q35: CANCEL: No found TPM device on host, output: ls: cannot access '/dev/tpm0': No such file or directory (2.65 s)

LGTM.

**Pull request merged**